safenode/src/protocol/online_transfers/mod.rs

```diff
@@ -0,0 +1,389 @@
+                match fee.content.decrypt_amount(&derived_key) {
+                    Ok(amount) => decrypted_node_fees.push(((node_id, fee), amount)),
+                    Err(error) => {
+                        error!("Decrypting the fee content from {node_id:?} failed! {error}");
+                        let _ = invalid_fees.insert(fee.content.reward_address);
+                    }
+                }
+            }
+
+            let max_invalid_fees = num_responses - required_responses;
+            if invalid_fees.len() > max_invalid_fees {
+                let valid_responses = num_responses - invalid_fees.len();
+                warn!("Not enough valid fees received from nodes to spend the input. Found: {valid_responses}, needed: {required_responses}", );
+                continue;
+            }
+
+            // Total fee paid to all recipients in the section for this input.
+            let fee_per_input = decrypted_node_fees
+                .iter()
+                .fold(Some(Token::zero()), |total, (_, fee)| {
+                    total.and_then(|t| t.checked_add(*fee))
+                })
+                .ok_or_else(|| Error::DbcReissueFailed(
+                    "Overflow occurred while summing the individual node's fees in order to calculate the total amount for the output DBCs."
+                        .to_string(),
+                ))?;
+
+            let mut fee_cipher_params = BTreeMap::new();
+
+            // Add nodes to outputs and generate their fee ciphers.
+            decrypted_node_fees
+                .iter()
+                .for_each(|((node_id, required_fee), fee)| {
+                    let dbc_id_src = required_fee
+                        .content
+                        .reward_address
+                        .random_dbc_id_src(&mut rand::thread_rng());
+                    recipients.push((*fee, dbc_id_src));
+                    let _ = fee_cipher_params.insert(*node_id, (required_fee.clone(), dbc_id_src));
+                });
+
+            let _ = all_fee_cipher_params.insert(dbc_id, fee_cipher_params);
+
+            fees_paid = fees_paid.checked_add(fee_per_input).ok_or_else(|| {
+                Error::DbcReissueFailed(
+                    "Overflow occurred while summing all the input fees.".to_string(),
+                )
+            })?;
+
+            fee_per_input
+        };
+        // ---------------- fee part end ----------------
+
+        // Add this dbc as input to be spent.
+        dbcs_to_spend.push((dbc, derived_key));
+
+        // Input amount increases with the amount of the dbc.
+        total_input_amount = total_input_amount.checked_add(dbc_balance)
+            .ok_or_else(|| {
+                Error::DbcReissueFailed(
+                    "Overflow occurred while increasing total input amount while trying to cover the output DBCs."
+                    .to_string(),
+            )
+            })?;
+
+        // ---------------- fee part start ----------------
```

Comment:
this part is about output, not `fee` ?

Reply:
> total_output_amount.checked_add(fee_per_input)

It is about fees.

Reply:
first, this `fee part` is duplicated with the above `fee part`
second, `total_output_amount.checked_add(fee_per_input)` is more about `output compose` instead of `fee calculation`
hence, maybe rename this part to be `output composing part` ?

Reply:
It is not duplicated. 

It's separated because this part was feature gated before (and might be still, we haven't decided on that yet).

This code part was only run when fees were used. Thus it is a `fee part`.

The other ones are about calculating total output _without_ fees.
But that is not important to distinguish, since that is never feature gated.

The fees though may be, therefore these parts which are specifically for fees, are marked here as `fee part start`and `fee part end`.